enhancements/network/static-ip-addresses-vsphere.md

```diff
@@ -0,0 +1,428 @@
+// IPAddressClaims for fulfillment by an external controller.
+type AddressesFromPool struct {
+	// APIGroup is the group for the resource
+	// being referenced. If APIGroup is not
+	// specified, the specified Kind must be in
+	// the core API group. For any other
+	// third-party types, APIGroup is required.
+	ApiGroup string `json:"apiGroup"`
+
+	// Kind is the type of resource being referenced
+	Kind string `json:"kind"`
+
+	// Name is the name of resource being referenced
+	Name string `json:"name"`
+}
+
+// Host defines the network device configuration to be applied for a node deployed by the installer
+type Host struct {  
+  // FailureDomain refers to the name of a FailureDomain as described in https://github.com/openshift/enhancements/blob/master/enhancements/installer/vsphere-ipi-zonal.md
+  // +optional
+  FailureDomain string `json: "failureDomain"`
+
+  // Slice of NetworkDeviceSpecs to be applied
+  // +kubebuilder:validation:Required
+  NetworkDevice []NetworkDeviceSpec `json: "networkDevice"` 
+
+  // Role defines the role of the node
+  // +kubebuilder:validation:Enum="";bootstrap;control-plane;compute
+  // +kubebuilder:validation:Required
+  Role string `json: "role"`
+}
+
+type NetworkDeviceSpec struct {
+	// gateway4 is the IPv4 gateway used by this device.
+	// Required when DHCP4 is false.
+	// +optional
+	// +kubebuilder:validation:Format=ipv4
+	Gateway4 string `json:"gateway4,omitempty"`
+
+	// gateway4 is the IPv4 gateway used by this device.
+	// Required when DHCP6 is false.
+	// +kubebuilder:validation:Format=ipv6
+	// +optional
+	Gateway6 string `json:"gateway6,omitempty"`
+
+	// ipaddrs is a list of one or more IPv4 and/or IPv6 addresses to assign
+	// to this device.
+	// Required when DHCP4 and DHCP6 are both Disabled.
+	// + Validation is applied via a patch, we validate the format as either ipv4 or ipv6
+	// +optional
+	IPAddrs []string `json:"ipAddrs,omitempty"`
+
+	// nameservers is a list of IPv4 and/or IPv6 addresses used as DNS
+	// nameservers.
+	// Please note that Linux allows only three nameservers (https://linux.die.net/man/5/resolv.conf).
+	// +optional
+	Nameservers []string `json:"nameservers,omitempty"`
+  
+  // addressesFromPools is a list of address pools which will fulfill requests for static IPs
+  AddressesFromPools []AddressesFromPool `json:"addressesFromPool,omitempty"`
+}
+
+~~~
+
+2. Add validation for the modified/added fields in the platform specification.
+3. For compute nodes, produce machine manifests with associated network device configuration.  
+4. For bootstrap and control plane nodes, modify vSphere terraform to convert network device configuration to a VM guestinfo parameter for each VM to be created.
+
+As the assets are generated for the control plane and compute nodes, the slice of `host`s for each node role will be used to populate network device configuration.  The number of `host`s must match the number of replicas defined in the associated machine pool.
```

Comment:
(This is as good a place as any for this question...)

How will the [MachineSets()](https://github.com/openshift/hive/blob/a9aaf72f8d1849aa5b5f2e3d6e137e93d9fbbdb5/vendor/github.com/openshift/installer/pkg/asset/machines/vsphere/machinesets.go#L133) function be impacted? Will the signature change? Will it accept the new things via the InstallConfig param? Something else?

Reply:
@2uasimojo I just saw this, my apologies!  The MachineSets function will not change.  When configured for static IPs, the machineset generated by the installer is set to 0 replicas, but is otherwise unchanged.  For static IPs, the installer will generate the initial set of compute resources.  For 4.14, this is going to be done under the scope of tech preview.

Reply:
Thanks @rvanderp3. I may need to dive a little deeper:
A common use case for hive users is to create a (hive) MachinePool alongside a ClusterDeployment to manage the number of replicas in the default worker pool. Do I understand correctly that the number of replicas will be fixed/immutable when static IPs are in play?

And to clarify
> the machineset generated by the installer is set to 0 replicas
The logic underneath MachineSets() will trigger a different code path based on seeing the static IP configuration in install-config such that [this](https://github.com/openshift/hive/blob/a9aaf72f8d1849aa5b5f2e3d6e137e93d9fbbdb5/vendor/github.com/openshift/installer/pkg/asset/machines/vsphere/machinesets.go#L177) will wind up being zero?

If I'm just way off, it may be easier for us to Meet™ and discuss; lmk.

Reply:
>  Do I understand correctly that the number of replicas will be fixed/immutable when static IPs are in play?

No, you can have any number of replicas.  By default, however, the installer takes care of assigning IPs to bootstrap the installation.  After installation, machinesets can be scaled by using an external IPAM controller.  I can see where this may present some challenges for you.  Is static IP support on the roadmap for you? 
p.s. apologies on it taking this long to get back to you